### PR TITLE
4133 WIP smallest change toward consolidated CI config

### PR DESCRIPTION
**Pull request by blaisep**
# WIP 4133 Towards a consolidated CI config
In today's session with @hacklschorsch and @crwood , we decided to fix the problem by making the smallest possible change.
Also, there is an outline of the path forward at the end of this note [Proposal: For release 1.20]

## Problem: Wheels fail to build,  when `pip install  git+https`

```shell
python -m venv venv && . venv/bin/activate
pip install git+https://github.com/tahoe-lafs/tahoe-lafs
```
fails with:

```
WARNING: Built wheel for tahoe-lafs is invalid: Wheel has unexpected file name: expected '0.0.0', got '1.19.0.post194'
Failed to build tahoe-lafs
ERROR: Could not build wheels for tahoe-lafs, which is required to install pyproject.toml-based projects
```
## Description
- setup.py has version logic that is not accessible from pyproject.toml
- "new tooling" (eg. `pip install` Tahoe from git.)
- new tooling does *NOT* run `setup.py update-version`
- new tooling does not get to use the version info
- modern tools prioritize

## Proposal: For release 1.20:
Make a smaller change: Choose one of the many bandaid approaches making setup.py use pyproject.toml (eg. versioningit or setuptools-scm)
- Now: patch pyproject.toml to include version info
- Next: move declarative items to pyproject.toml
- Later: 
  - change `setuptools` to take version info from pyproject.toml
  - Deprecate `setup.py`

## Proposal for >1.20: adopt Hatch

Edit `pyproject.toml`
to include"
```
# If using Hatch:
[build-system]
requires = [
    "hatchling",
    "versioningit",
]
build-backend = "hatchling.build"

# This setting is also required if you're using Hatch:
[tool.hatch.version]
source = "versioningit"
```

## References

- https://packaging.python.org/en/latest/key_projects/
- https://github.com/jwodder/versioningit?tab=readme-ov-file#installation--setup
- https://github.com/jwodder/versioningit
- https://github.com/pypa/setuptools-scm?tab=readme-ov-file#pyprojecttoml-usage

Co-authored-by: Jeff Jacobson <jeffjacobsonhimself@gmail.com>

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -130,7 +130,7 @@ workflows:
               # of Python, we probably have to do something here.
               pythonVersion:
                 - "3.9"
-                - "3.11"
+                - "3.12"
```
Comment (hacklschorsch) [inline .circleci/config.yml:133]: I am confused.  Aren't all these changes already in `master`?

Comment (meejah) [reply .circleci/config.yml:133]: The "sych with Upstream" commit smashes a bunch together. This is not the way to re-sync (so this PR will be un-mergable). We can discuss in N&B tomorrow.

The tl;dr is: "git merge master" to resync a branch.

Comment (hacklschorsch) [reply .circleci/config.yml:133]: This PR is building on top of the PR #1381 I use to poke around with CI?
That is a misunderstanding, I do not intend to merge that branch - Anything good in there I have created small new PRs for.  I really only use it as a sandbox.  Please don't build on top of it.

Comment (hacklschorsch) [reply .circleci/config.yml:133]: > The "sych with Upstream" commit smashes a bunch together. This is not the way to re-sync

I didn't mean to merge master in, I intended to throw away the prior work in that PR to start over again with what was on master at that time.

Comment (blaisep) [reply .circleci/config.yml:133]: Sorry, I will close this PR.

**Pull request closed without merging**